### PR TITLE
New: Support Definite Assignment (fixes #424)

### LINKED ISSUE #424: Definite assignment assertions not in AST
**What version of TypeScript are you using?**
` typescript@2.7.0-dev.20171224`

**What version of `typescript-eslint-parser` are you using?**
`master`

**What code were you trying to parse?**

```ts
class X { a!: any }
```

**What did you expect to happen?**
Something like `definite: true` or `exclamationToken: true` in the AST. In the TS AST it is called `exclamationToken`.

New in https://github.com/Microsoft/TypeScript/pull/20166

**What happened?**

```js
{
  type: 'ClassProperty',
  range: [ 10, 17 ],
  loc: {...},
  key:  {...},
  value: null,
  computed: false,
  static: false,
  readonly: undefined,
  typeAnnotation:  {...}
}
```

Thoughts on which property name to go for?

### PR DESCRIPTION
**Pull request by azz**
Build will fail due to this not yet supported in Babylon.

### CODE REVIEW

**Comment by azz** [comment]
Did you catch this comment? https://github.com/eslint/typescript-eslint-parser/issues/424#issuecomment-356285051

**Comment by JamesHenry** [comment]
Added my 2 cents :)

**Comment by azz** [comment]
I have no strong opinions either way 🤷‍♂️

**Comment by JamesHenry** [comment]
As with the unique symbol one, I've started a protected `ts-2.7` branch as we normally do with changes which relate to not yet stable versions of TS. Please change the base to be that branch

**Comment by JamesHenry** [comment]
I'm the same regarding opinion of the change - but I am strongly of the opinion that we choose whatever ends up being used in that PR so that we stay aligned 👍

**Comment by JamesHenry** [comment]
Again, with this one @azz I am just going to merge this into the 2.7 branch and then sort out any issues/cleanup. Thanks again!

**Comment by JamesHenry** [comment]
Ah no, I can't sorry, cause of the conflicts. There is also yarn.lock file that needs to be removed.

**Comment by JamesHenry** [comment]
I've pressed ahead with the TS 2.7 release so that's all merged into master now.

**Comment by azz** [comment]
Changed base branch and fixed conflicts.

**Comment by azz** [comment]
Ping @JamesHenry, we'd really like to get this out to Prettier users!

File: lib/convert.js [hunk 1]
```diff
@@ -859,6 +863,10 @@ module.exports = function convert(config) {
                 result.key.optional = true;
             }
 
+            if (node.exclamationToken) {
+                result.exclamationToken = true;
```
Comment (j-f1) [inline lib/convert.js]: I [suggested](https://github.com/babel/babel/pull/7159#discussion_r167437266) `definite` in the Babylon PR, what do you think?

Comment (azz) [reply lib/convert.js]: We'll follow babylon's choice - updated to `definite`.

**Comment by JamesHenry** [comment]
The build isn't passing, please just add your new specs to the ignore patterns within ast-alignment testing for now: https://github.com/eslint/typescript-eslint-parser/blob/master/tests/ast-alignment/fixtures-to-test.js#L341

File: lib/convert.js (hunk 1 above)
Comment (JamesHenry) [reply lib/convert.js]: Thanks 👍

**Comment by azz** [comment]
Ah, sorry! Updated.

**Comment by JamesHenry** [comment]
Thanks, @azz!

**Comment by JamesHenry** [comment]
Released in 14.0.0

**Pull request merged**